core/modules/block/block.module

```diff
@@ -234,8 +303,34 @@ function block_custom_block_form($edit = array(), $stand_alone = TRUE) {
     '#weight' => -16,
   );
 
+  // Add default language select dropdown on the stand-alone config form only.
+  // Not shown on the translation forms at all.
+  if ($stand_alone && module_exists('locale') && empty($edit['langcode'])) {
+    $language_options = array(LANGUAGE_NONE => t('- None -'));
+    $language_options += language_list(TRUE, TRUE);
+
+    $form['default_langcode'] = array(
+      '#type' => 'select',
+      '#title' => t('Language'),
+      '#description' => 'If assigned a language, options will be shown to translate this block into other langauges.',
+      '#options' => $language_options,
+      '#default_value' => $edit['default_langcode'],
+    );
+  }
+
   // When displaying as part of the Layout UI.
   if (!$stand_alone) {
+    // The Layout UI may show the translated block admin title and description,
+    // but this form will only show the source translation contents. Notify
+    // the user that they may be about to edit something they weren't expecting.
+    if (!empty($edit['translations'])) {
+      $form['translation_help'] = array(
+        '#type' => 'help',
+        '#markup' => '<strong>' . t('Translation notice') . ':</strong> ' . t('The settings in this form are for the source translation only. Translations may be modified on the <a href="!url">block\'s translation settings page</a>.', array('!url' => url('admin/structure/block/' . $edit['delta'] . '/translation'))),
```

Comment:
The URL here needs to be like this `admin/structure/block/manage/*/translation`, currently it's missing the "manage".